server/s3_list_object_versions_action.cc

```diff
@@ -0,0 +1,835 @@
+        request->set_out_header_value("Retry-After",
+                                      std::to_string(retry_after_period));
+      } else {
+        request->set_out_header_value("Retry-After", "1");
+      }
+    }
+    request->send_response(error.get_http_status_code(), response_xml);
+  } else if (fetch_successful) {
+    std::string& response_xml = get_response_xml();
+    request->set_out_header_value("Content-Length",
+                                  std::to_string(response_xml.length()));
+    request->set_bytes_sent(response_xml.length());
+    request->set_out_header_value("Content-Type", "application/xml");
+    s3_log(S3_LOG_DEBUG, request_id, "Object list response_xml = %s\n",
+           response_xml.c_str());
+    // Total visited/touched keys in the bucket
+    s3_log(S3_LOG_INFO, stripped_request_id, "Total keys visited = %zu\n",
+           total_keys_visited);
+    request->send_response(S3HttpSuccess200, response_xml);
+  } else {
+    S3Error error("InternalError", request->get_request_id(),
+                  request->get_bucket_name());
+    std::string& response_xml = error.to_xml();
+    request->set_out_header_value("Content-Type", "application/xml");
+    request->set_out_header_value("Content-Length",
+                                  std::to_string(response_xml.length()));
+
+    request->send_response(error.get_http_status_code(), response_xml);
+  }
+  S3_RESET_SHUTDOWN_SIGNAL;  // for shutdown testcases
+  done();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3ListObjectVersionsAction::add_marker_version(std::string key,
+                                                    std::string version_id) {
+  key_marker = key;
+  version_id_marker = version_id;
+}
+
+void S3ListObjectVersionsAction::add_next_marker_version(
+    std::string next_key, std::string next_value) {
+  m0_uint128 object_version_list_index_oid =
+      bucket_metadata->get_objects_version_list_index_layout().oid;
+  auto next_version =
+      object_metadata_factory->create_object_metadata_obj(request);
+  if (next_version->from_json(next_value) != 0) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Json Parsing failed. Index oid = "
+           "%" SCNx64 " : %" SCNx64 ", Key = %s, Value = %s\n",
+           object_version_list_index_oid.u_hi,
+           object_version_list_index_oid.u_lo, next_key.c_str(),
+           next_value.c_str());
+  } else {
+    next_key_marker = next_version->get_object_name();
+    next_version_id_marker = next_version->get_obj_version_id();
+  }
+}
+
+void S3ListObjectVersionsAction::add_object_version(std::string key,
+                                                    std::string value) {
+  m0_uint128 object_version_list_index_oid =
+      bucket_metadata->get_objects_version_list_index_layout().oid;
+  auto version = object_metadata_factory->create_object_metadata_obj(request);
+  if (version->from_json(value) != 0) {
+    json_error = true;
```

Comment:
Will be handled in https://jts.seagate.com/browse/EOS-26911

Reply:
.

Reply:
It will be difficult to change it at this stage as by design function `add_object_version()` is called at multiple places within the `if` block and the error is checked only once, outside of the `if` block. We should leave it as is.